### DISCUSSION TITLE
Custom Resources - How to define & pass ports?

### DISCUSSION THREAD

**Discussion by jakenuts** [microsoft/aspire#1373] [Q&A]
I'm trying to create a resource for the Temporal.io dev server which launches an backend server and a monitoring ui. 
I'm at the point where the server is launched, the url for the UI is available in the Aspire dashboard, but I can't determine the proper way to pass the server endpoint to consuming projects or why the default ports I'm providing (7233, 8233) are being ignored. As the temporal client doesn't accept an HttpClient (it's grpc?) the usual samples don't seem to fit.

Any ideas what's wrong/right here? 

Thanks!! 8)

The extension method

![image](https://github.com/dotnet/aspire/assets/344606/482ecde6-033f-4ca8-83ed-b207b2062cde)

The AppHost setup

![image](https://github.com/dotnet/aspire/assets/344606/a2e7ee90-63c4-4a5c-afef-af10d9464c0d)

The resource (prolly don't need the connection string bit)

![image](https://github.com/dotnet/aspire/assets/344606/2e298301-1a1e-4ab5-b9b5-aee9ba611091)

When I run this I get these in the dependant apps

Web

![image](https://github.com/dotnet/aspire/assets/344606/c8c05a08-fd32-4f2a-8304-b2de84f1426f)

Worker

![image](https://github.com/dotnet/aspire/assets/344606/5f724428-1f31-424a-a93d-7aef895daa21)

Aspire Dashboard

![image](https://github.com/dotnet/aspire/assets/344606/3183b26c-2eab-4839-a384-b30071573f46)

**Comment by DamianEdwards** [comment]
gRPC is still http from a service binding / service discovery point of view, so you've got a resource that exposes two http endpoints. Because the scheme ("http") is used as the endpoint name by default, you'll need to disambiguate them by naming the endpoints. You can keep one as just "http" but the other will need an explicit name. I suggest keeping the gRPC endpoint as the default "http" endpoint so that consuming apps don't need to specify the endpoint name to resolve "http://temporal-server", and explicitly name the UI endpoint as "ui" or similar.

**Comment by DamianEdwards** [comment]
Also see https://learn.microsoft.com/dotnet/aspire/service-discovery/overview for more information but this document needs some updates to better cover your scenario.

**Comment by jakenuts** [comment]
Thanks! 

Two quick questions, in the code below, 

1. Is WithEnvironment the recommended way of passing the endpoint a dependant project?
2. Should I expect the GetEndpoint("http") to be http://localhost:7233 if that was free or will it always generate a dynamic port and if so, how could I reference that dynamic port when launching the development server?


```
var temporalBuilder = builder.AddResource(new TemporalDevelopmentServerResource(...))
            .WithServiceBinding(7233, "http")         // The gRPC server

 var workerservice = builder.AddProject<DealerVision_Temporal_WorkerService>(AspireService.WorkerService)
       .WithEnvironment("temporal_server", temporal.GetEndpoint("http"));

Debug.Writeline(temporal.GetEndpoint("http"));

> "http://localhost:56905"
```

Thanks again, I know this is probably somewhere in the documentation but I'm a bit slow. 8)

**Reply by DamianEdwards** [reply]
For projects you should be able to use `WithReference` to pass in the gRPC server resource and that will result in environment variables being created that contribute to the configuration of service discovery in the project. The format of the names for those environment variables can't be changed though.

A proxy is stood-up in front of all resources currently, so that's likely where the dynamic port is coming from. The behavior of this is changing in preview.2 so that you should see the port you specified be the port used by the proxy and thus returned by the `GetEndpoint` call in preview.2.

**Reply by jakenuts** [reply]
Thanks!

**Comment by davidfowl** [comment]
As a follow up to this, here's the doc that describes how networking works https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/networking-overview

**Comment by ElanHasson** [comment]
@jakenuts is this something you've published a nuget for? I was just about to make this :)

**Comment by davidfowl** [comment]
I wrote a doc on the resource model. Let me know if it helps https://gist.github.com/davidfowl/b408af870d4b5b54a28bf18bffa127e1
